File: cpp/cpp_100-999/241.cc
// C++
// leetcode 241
// https://leetcode.cn/problems/different-ways-to-add-parentheses/

class Solution {
public:
    vector<int> diffWaysToCompute(string expression) {
    	return dfs(expression, 0, expression.size() - 1);
    }

    vector<int> dfs(string &s, int l, int r) {
    	vector<int> res;
    	for(int i = l; i <= r; i++) {
    		if(isNumChar(s[i]))
    			continue;
    		vector<int> left = dfs(s, l, i-1);
    		vector<int> right = dfs(s, i + 1, r);
    		for(auto x: left) {
    			for(auto y : right) {
    				if(s[i] == '+') {
    					res.push_back(x+y);
    				} else if (s[i] == '-'){
    					res.push_back(x-y);
    				} else {
    					res.push_back(x*y);
    				}
    			}
    		}
    	}
    	if (res.empty()) {
    		res.push_back(stoi(s.substr(l, r-l+1)));
    	}
    	return res;
    }

    bool isNumChar(char c) {
    	if(c >= '0' && c <= '9')
    		return true;
    	return false;
    }
};

File: Go/Go_100-999/241.go
// Go
// leetcode 241
// https://leetcode.cn/problems/different-ways-to-add-parentheses/

func diffWaysToCompute(expression string) []int {
	return dfs(expression, 0, len(expression) - 1)
}

func dfs(s string, l, r int) (res []int) {
	for i:= l; i <= r; i++ {

		if s[i] >= '0' && s[i] <= '9' {
			continue
		}

		left := dfs(s, l, i - 1)
		right := dfs(s, i + 1, r)
		for _, x := range left {
			for _, y := range right {
				if s[i] == '+' {
					res = append(res, x+y)
				} else if s[i] == '-' {
					res = append(res, x-y)
				} else {
					res = append(res, x*y)
				}
			}
		}
	}

	if len(res) == 0 {
		tmp := 0
		for i := l; i <= r; i++ {
			tmp = tmp * 10 + int(s[i] - '0')
		}
		return []int{tmp}
	}

	return res
}